More projectile property streamlining and cleanups

reaperrr opened the pull request:
Another small step towards more consistency in property names, also some negligible performance tweaks and Lint rules to warn modders about invalid launch angles.

All yaml changes were performed with the respective upgrade rules.

reaperrr committed:
Cache Missile speed infos in constructor

The primary reason is to have one central location for the look-ups, so we'll only have to change up to 3 lines if we rename or merge the speed properties.

reaperrr committed:
Cache Missile LaunchAngles in constructor

To have one central place for the info look-ups. Makes it a little easier to rename the properties or merge them into a single Angle[] property.

reaperrr committed:
Rename Bullet Angle to LaunchAngle

reaperrr commented:
> OpenRA.Utility(1,1): Error: OpenRA.Mods.Common.Lint.CheckAngle failed with exception: System.NullReferenceException: Object reference not set to an instance of an object

Does anyone have an idea what the hell is going on here?
Because this test *passes* on my local branch (I didn't make any changes since I pushed).

pchote commented on OpenRA.Mods.Common/Projectiles/AreaBeam.cs:
```diff
@@ -28,7 +28,7 @@ public class AreaBeamInfo : IProjectileInfo
 
 		[Desc("The maximum duration (in ticks) of each beam burst.")]
-		public readonly int Duration = 10;
+		public readonly int BeamDuration = 10;
```
We already know that this is a beam, so adding it again here doesn't help.

reaperrr replied:
I intend to add `DamageDuration` in my next PR and felt this would be better to separate them, but if you prefer to rename `BeamDuration` in `LaserZap` to just `Duration` instead, I guess I can do that.

pchote commented on OpenRA.Mods.Common/Projectiles/LaserZap.cs:
```diff
@@ -23,6 +23,9 @@ namespace OpenRA.Mods.Common.Projectiles
 	public class LaserZapInfo : IProjectileInfo
 	{
+		[Desc("Should the beam be visually rendered? False = Beam is invisible.")]
+		public readonly bool RenderBeam = true;
```
A laser that isn't rendered isn't a laser at all (just a normal projectile), so isn't suited for this projectile type at all.

pchote commented on OpenRA.Mods.Common/Projectiles/LaserZap.cs:
```diff
@@ -100,17 +103,18 @@ public void Tick(World world)
-			if (ticks < info.BeamDuration)
+			var targetObscured = wr.World.FogObscures(target);
+			var sourceObscured = wr.World.FogObscures(args.Source);
+			if (!(targetObscured && sourceObscured) && ticks < info.BeamDuration)
```
This will cause the whole beam to be invisible if you get shot by an obelisk hidden under the fog, which looks much worse than what we have right now (I had already tried this a long time ago).

reaperrr commented:
Updated.

However, I still need help with the lint check. For some reason it only works with positive values, negative angles always trigger the lint error (which is why I've disabled it for missiles for now).

pchote commented:
`WAngle` is only defined for numbers between 0 and 1024.  The `new WAngle(-64)` in Missile.cs is internally converted to 960, which then fails your > 255 check.

reaperrr commented:
> You check should be something like `return value > 255 && value < 768`;

Done, though I went for `value < 76*9*` because 768 should be the highest invalid value (1024 - 256 = 768).

GraionDilach commented on OpenRA.Mods.Common/Projectiles/Missile.cs:
```diff
@@ -37,20 +37,12 @@ public class MissileInfo : IProjectileInfo
-		public readonly WDist Speed = new WDist(384);
+		[Desc("Projectile speed in WDist / tick, two values indicate launch speed and maximum speed,",
+			"three values represent minimum launch speed, maximum launch speed and maximum speed (in that order).")]
+		public readonly WDist[] Speed = { new WDist(384) };
```
Notcertain if I'm a fan of this, this forces unneeded duplication at inherited ones.

reaperrr commented:
Removed the first two laser-related commits, which are included in #12068 now.

abcdefg30 commented:
Needs a rebase.

reaperrr commented:
Before I rebase this, any other opinions on the angle and speed field merging?

I see @GraionDilach's point, so I could live with dropping the commits in question.

pchote commented:
I agree with @GraionDilach, so lets drop those commits.

reaperrr committed:
Add lint check for invalid projectile angles

Emits error if
- angle exceeds 255 or -255
- minimum launch angle exceeds maximum launch angle

reaperrr commented:
Updated. 
Removed the commits in question and extended lint check to also emit errors when minimum launch angle exceeds maximum launch angle.

reaperrr committed:
Sync less bullet aspects

`length` is calculated from `pos` and `target` which are both synced.
`SourceActor`, `ticks` and `smokeTicks` work no different from `Missile`, where they're not synced, either.

reaperrr commented:
Updated.

Removed some syncing from `Bullet`, as it seemed excessive compared to `Missile`.

pchote reviewed (commented):
Looks sensible overall, but I haven't looked in full detail or tested, so :+0.5: for now.

pchote commented on OpenRA.Mods.Common/Lint/CheckAngle.cs:
```diff
@@ -0,0 +1,67 @@
+			foreach (var weaponInfo in rules.Weapons)
+			{
+				var missile = weaponInfo.Value.Projectile as MissileInfo;
+				if (missile != null)
```
Can you please pull this out into a `void CheckLaunchAngles(WAngle min, WAngle max)` to avoid duplication between these cases?

pchote reviewed (dismissed):
Looks good after this fix

reaperrr commented:
Updated the lint check.

Also made it slightly smarter, `maxAngle` is now only tested if it's not identical to `minAngle` to avoid duplicate errors when that single/identical angle is invalid.

abcdefg30 commented:
Looks good to me. :+1: / :white_check_mark:

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/7f462506ec75487de697078778867d61862f5891
